src/builder.rs

```diff
@@ -165,26 +166,21 @@ impl SessionBuilder {
         // if the call _didn't_ error, then the backgrounded ssh client will still hold onto those
         // handles, and it's still running, so those reads will hang indefinitely.
         let mut child = init.spawn().map_err(Error::Connect)?;
-        let status = child.wait().map_err(Error::Connect)?;
-
-        if let Some(255) = status.code() {
-            // this is the ssh command's way of telling us that the connection failed
-            let mut stderr = String::new();
-            child
-                .stderr
-                .as_mut()
-                .unwrap()
-                .read_to_string(&mut stderr)
-                .unwrap();
-
-            return Err(Error::interpret_ssh_error(&stderr));
+        let stdout = child.stdout.take().unwrap();
+        let mut stderr = child.stderr.take().unwrap();
+        let status = child.await.map_err(Error::Connect)?;
+
+        if !status.success() {
+            let mut err = String::new();
+            stderr.read_to_string(&mut err).await.unwrap();
+            return Err(Error::interpret_ssh_error(&err));
         }
 
         Ok(Session {
             ctl: dir,
             addr: String::from(destination),
             terminated: false,
-            master: std::sync::Mutex::new(Some(child)),
+            master: std::sync::Mutex::new(Some((stdout, stderr))),
```

Comment:
Because we've already waited for the `child`, and we know it didn't fail, so keeping it around doesn't actually make a difference. All we care about is the output channels, since it will have passed those to the backgrounded master connection process.